File: src/main/java/org/arbor/gtnn/GTNNRegistries.java
package org.arbor.gtnn;

import com.gregtechceu.gtceu.api.GTCEuAPI;
import com.gregtechceu.gtceu.api.data.chemical.material.event.MaterialEvent;
import com.gregtechceu.gtceu.api.data.chemical.material.event.MaterialRegistryEvent;
import com.gregtechceu.gtceu.api.data.chemical.material.registry.MaterialRegistry;
import com.gregtechceu.gtceu.api.machine.MachineDefinition;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.packs.FilePackResources;
import net.minecraft.server.packs.PackResources;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;
import org.apache.commons.io.FileUtils;
import org.arbor.gtnn.api.registry.GTNNRegistrate;
import org.arbor.gtnn.data.GTNNMachines;
import org.arbor.gtnn.data.GTNNMaterials;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;

@Mod.EventBusSubscriber(modid = GTNN.MODID, bus = Mod.EventBusSubscriber.Bus.MOD)
public final class GTNNRegistries {
    public static final GTNNRegistrate REGISTRATE = GTNNRegistrate.create(GTNN.MODID);
    public static MaterialRegistry MATERIAL_REGISTRY;

    public static PackResources getAllPackResources(){
        File tempFile = null;
        try {
            InputStream inputStream = GTNNRegistries.class.getResourceAsStream("/data/gtnn/ad_astra.zip");
            tempFile = File.createTempFile("temp", ".tmp");
            assert inputStream != null;
            FileUtils.copyInputStreamToFile(inputStream, tempFile);
            inputStream.close();
        } catch (IOException e) {
            GTNN.LOGGER.error("getAllPackResources() wrong!", e);
        }
        if (tempFile != null) {
            return new FilePackResources(tempFile.getPath(), tempFile, false);
        }
        return null;
    }

    public static void registerMachine(GTCEuAPI.RegisterEvent<ResourceLocation, MachineDefinition> ignoredEvent){
        GTNNMachines.init();
    }

    @SubscribeEvent
    public static void registerMaterialRegistryEvent(MaterialRegistryEvent event) {
        MATERIAL_REGISTRY = GTCEuAPI.materialManager.createRegistry(GTNN.MODID);
    }

    @SubscribeEvent
    public static void registerMaterials(MaterialEvent event) {
        GTNNMaterials.init();
    }

    public GTNNRegistries() {
    }
}

File: src/main/kotlin/dev/arbor/gtnn/GTNNRegistries.kt
package dev.arbor.gtnn

import com.gregtechceu.gtceu.api.GTCEuAPI
import com.gregtechceu.gtceu.api.data.chemical.material.event.MaterialEvent
import com.gregtechceu.gtceu.api.data.chemical.material.event.MaterialRegistryEvent
import com.gregtechceu.gtceu.api.data.chemical.material.registry.MaterialRegistry
import com.gregtechceu.gtceu.api.machine.MachineDefinition
import com.gregtechceu.gtceu.api.registry.registrate.GTRegistrate
import dev.arbor.gtnn.api.block.NNBlockMaps
import dev.arbor.gtnn.api.lang.LangGenerators
import dev.arbor.gtnn.api.registry.GTRecipeEvent
import dev.arbor.gtnn.api.registry.GTRecipeManager
import dev.arbor.gtnn.api.registry.NNRegistrate
import dev.arbor.gtnn.client.ExtraHeartRenderHandler
import dev.arbor.gtnn.client.GetRecipesCommand.register
import dev.arbor.gtnn.client.StructureSelectRenderer
import dev.arbor.gtnn.client.renderer.item.ItemCustomLayerModel
import dev.arbor.gtnn.data.GTNNMachines
import dev.arbor.gtnn.data.GTNNMaterials
import dev.arbor.gtnn.data.GTNNRecipes
import dev.arbor.gtnn.data.lang.MachineLang
import dev.arbor.gtnn.temp.VirtualEnderRegistry
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.packs.FilePackResources
import net.minecraft.server.packs.PackResources
import net.minecraftforge.client.event.ModelEvent.RegisterGeometryLoaders
import net.minecraftforge.client.event.RenderLevelStageEvent
import net.minecraftforge.client.event.RenderLevelStageEvent.Stage
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.event.RegisterCommandsEvent
import net.minecraftforge.event.server.ServerStoppedEvent
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent
import org.apache.commons.io.FileUtils
import java.io.File

@Suppress("unused")
object GTNNRegistries {
    private lateinit var MATERIAL_REGISTRY: MaterialRegistry

    val REGISTRATE: NNRegistrate by lazy {
        NNRegistrate(GTNN.MOD_ID).also(::addAdditionalDataGenerators)
    }

    @JvmStatic
    fun registerRecipeHandler(event: GTRecipeEvent.RegisterHandler) {
        GTNNRecipes.register(event)
    }

    @JvmStatic
    fun registerGeometryLoaders(event: RegisterGeometryLoaders) {
        event.register("item_custom_layers", ItemCustomLayerModel.INSTANCE)
    }

    @JvmStatic
    fun serverStopped(event: ServerStoppedEvent) {
        VirtualEnderRegistry.release()
    }

    @JvmStatic
    fun fmlCommonSetupEvent(event: FMLCommonSetupEvent) {
        MinecraftForge.EVENT_BUS.register(ExtraHeartRenderHandler())
        GTRecipeManager.onCommonSetup()
    }

    @JvmStatic
    fun onRenderWorldLast(event: RenderLevelStageEvent) {
        val stage = event.stage
        if (stage === Stage.AFTER_TRIPWIRE_BLOCKS) {
            StructureSelectRenderer.renderStructureSelect(event.poseStack, event.camera)
        }
    }

    @JvmStatic
    fun onServerStarting(event: RegisterCommandsEvent) {
        register(event.dispatcher)
    }

    @JvmStatic
    fun registerMachines(event: GTCEuAPI.RegisterEvent<ResourceLocation, MachineDefinition>) {
        GTNNMachines.init()
        GTNN.LOGGER.info("register GTNN Machines")
    }

    @JvmStatic
    fun onCommonSetup(modBus: FMLCommonSetupEvent) {
        NNBlockMaps.initBlocks()
    }

    @JvmStatic
    fun registerMaterialRegistryEvent(event: MaterialRegistryEvent) {
        MATERIAL_REGISTRY = GTCEuAPI.materialManager.createRegistry(GTNN.MOD_ID)
        GTNN.LOGGER.info("register GTNN Materials")
    }

    @JvmStatic
    fun registerMaterials(event: MaterialEvent) {
        GTNNMaterials.init()
    }

    @JvmStatic
    fun getAllPackResources(): List<PackResources> {
        val packResources = ArrayList<PackResources>()
        if (GTNNIntegration.isAdAstraLoaded()) {
            this::class.java.getResourceAsStream("/data/gtnn/ad_astra.zip").use {
                val tempFile = File.createTempFile("gtnn_resource_pack", ".tmp")
                FileUtils.copyInputStreamToFile(it, tempFile)
                packResources.add(FilePackResources(tempFile.getName(), tempFile, false))
            }
        }
        return packResources
    }

    private fun addAdditionalDataGenerators(registrate: GTRegistrate) {
        LangGenerators.initDatagen(registrate, MachineLang::class)
    }
}
